### PR TITLE
feat: separate settings for thumbnail and original photo storage locations in .env file #418

### LINKED ISSUE #418: [Feature]: Separate settings for Thumbnail and Original photo storage locations in .env file
### Feature detail

Separate storage path specification in .env for where to store thumbnails (fast ssd or other local storage) and original photos (slow/less expensive storage or cloud drive).

### Platform

Server

### PR DESCRIPTION
**Pull request by panoti**
Separate settings for thumbnail and original photo storage locations in .env file #418

- [x] use `UPLOAD_LOCATION` to define original photo and video storage location.
- [x] use `CACHE_LOCATION` to define thumbnail photo and encoded video storage location.

### CODE REVIEW

**Comment by bo0tzz** [comment]
Good stuff! I do think it would be nice to maintain the current default location in case the user doesn't set anything.

**Comment by panoti** [comment]
> Good stuff! I do think it would be nice to maintain the current default location in case the user doesn't set anything.

I forgot default value for `basePath`. I will add it soon.

**Comment by alextran1502** [comment]
Hi Thanh, I just have a look and ran some testing for this PR.

This implementation will break the current setup for everyone. Looks like you are implementing the way of not using Docker at all and running the server as a standalone application. This is not intended for running the app since we have many moving parts, so I discourage you from breaking away from it.


image

The `APP_UPLOAD_LOCATION` location is hardcoded in the code because it is mapped to the mount volume `UPLOAD_LOCATION` in the `.env` file. So if you change that value, the docker container doesn't know the context outside of its environment. Therefore, it cannot find the associated directory to upload the file.

**Review by alextran1502 — CHANGES_REQUESTED** [review]
Although, the intention of not using the Docker environment in running the app is not encouraged. There is a lot of quality change for the code base. 

Thank you.

**Comment by alextran1502** [comment]
> > Although, the intention of not using the Docker environment in running the app is not encouraged. There is a lot of quality change for the code base.
> > Thank you.
> 
> I run on Docker for testing too. I think user can change mount upload dir in different way without affecting code base. If they leave blank config, upload dir with be mounted with default config.

For my current setup, below is my observation.

1. I have `UPLOAD_LOCATION` specified as, let's say, `/home/alex/immich-data`
2. From `this.config.get('UPLOAD_LOCATION', APP_UPLOAD_LOCATION)`, the value will be `/home/alex/immich-data` but in the `immich-server`'s container context, it doesn't see the `/home/alex/immich-data` thus it cannot write to it.

The `APP_UPLOAD_LOCATION` is hardcoded as the path created by the server, so there will always be that path inside the container.

The reason you don't see we are using the `UPLOAD_LOCATION` value directly in the application is that it is there for the `docker-compose.yml` to map the directory to the predefined path in the container, see below.

image

**Comment by panoti** [comment]
@alextran1502 That's not problem

```yml
  volumes:
    - ${UPLOAD_LOCATION}:/usr/src/app/upload
```

`UPLOAD_LOCATION` variable is different from `UPLOAD_LOCATION` in `.env`. If you config `.env` with `UPLOAD_LOCATION`, for example:

```
# Upload File Config
UPLOAD_LOCATION=/uploads
``` 

You have to mount your volume with `/uploads`

```yml
  volumes:
    - ${UPLOAD_LOCATION}:/uploads
```

`UPLOAD_LOCATION` in volumes section is belong to host env. `UPLOAD_LOCATION` in `.env` is belong to container. Some end-users not care about `UPLOAD_LOCATION` where actually is, they just leave it with blank:

```
# Upload File Config
UPLOAD_LOCATION=
```

**Comment by alextran1502** [comment]
> Some end-users not care about UPLOAD_LOCATION where actually is, they just leave it with blank:

This is not correct. Immich is designed to mount your local directory to the container so that once the photos/videos are uploaded, the user will be able to access the file in the directory. the value of `UPLOAD_LOCATION` is not specified, and the user will have no way to access the files besides the web and the mobile app.

I hope this piece of information gives you an insight into the motivation and the intended structure of the app

**Comment by panoti** [comment]
Hey, you may be misunderstand me

```yml

services:
  immich-server:
    image: ghcr.io/panoti/immich/immich-server:latest
    entrypoint: ["/bin/sh", "./start-server.sh"]
    volumes:
      - immich_upload:/usr/src/app/upload
    environment:
      - NODE_ENV=production
      - DB_HOSTNAME=tasks.postgresdb
      - DB_USERNAME=username
      - DB_PASSWORD=password
      - DB_DATABASE_NAME=immich
      - REDIS_HOSTNAME=tasks.redis
      - ENABLE_MAPBOX=false
      - MAPBOX_KEY=
      - VITE_LOGIN_PAGE_MESSAGE=
```

We can mount local directory to the container `/usr/src/app/upload` path. As you can see, if I remove both `UPLOAD_LOCATION` and `THUMBNAIL_LOCATION`, we still mount to `/usr/src/app/upload` without changes.

**Comment by panoti** [comment]
Making use of `UPLOAD_LOCATION` variable is not feasible 😭. So I use other variables without making any current config changes. Now, both docker and pm2 users may be happy with flexible configs, I hope so 😿

File: server/apps/microservices/src/config/app.config.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { ConfigModuleOptions } from '@nestjs/config';
```
Comment (alextran1502) [inline server/apps/microservices/src/config/app.config.ts:1]: having two separate files doing the same thing is not a good thing. We can either move this into the `server/libs` directory where both `server` and `microservices` can share code.

Comment (panoti) [reply server/apps/microservices/src/config/app.config.ts:1]: I will store `app.config.ts` to `server/libs/config` and import again.

Comment (panoti) [reply server/apps/microservices/src/config/app.config.ts:1]: Do you wanna add alias `@app/config` in this pull? or I will import relative path like `../../../libs/config/app.config`

**Comment by alextran1502** [comment]
> Making use of `UPLOAD_LOCATION` variable is not feasible sob. So I use other variables without making any current config changes. Now, both docker and pm2/systemd users may be happy with flexible configs, I hope so crying_cat_face I begin my devops lessons from pm2 and systemd, so I wanna make every devops happy with current configssmile Document will be updated when this pull request is accepted

Please also update the proposed documentation change. I would like to see if it makes sense for the users.

**Comment by panoti** [comment]
> Please also update the proposed documentation change. I would like to see if it makes sense for the users.

I added some guide lines at README.md

File: server/apps/immich/src/api-v1/server-info/server-info.service.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import diskusage from 'diskusage';
 @Injectable()
 export class ServerInfoService {
   async getServerInfo(): Promise<ServerInfoResponseDto> {
-    const diskInfo = await diskusage.check(APP_UPLOAD_LOCATION);
+    const diskInfo = await diskusage.check(process.env.UPLOAD_DIR || APP_UPLOAD_LOCATION);
```
Comment (alextran1502) [inline server/apps/immich/src/api-v1/server-info/server-info.service.ts:9]: I don't see the `UPLOAD_DIR` in the `.env.example`. Is this a left over from previous changes?

Comment (alextran1502) [reply server/apps/immich/src/api-v1/server-info/server-info.service.ts:9]: Or is this for running the app as a standalone?

Comment (panoti) [reply server/apps/immich/src/api-v1/server-info/server-info.service.ts:9]: I wanna hide this env for docker user now. It's really useful with dockerless usage by using pm2, systemd, openrc, ...

Comment (alextran1502) [reply server/apps/immich/src/api-v1/server-info/server-info.service.ts:9]: Maybe there is a better way to implement this with documentation 🤔

Comment (panoti) [reply server/apps/immich/src/api-v1/server-info/server-info.service.ts:9]: I just start my first step.  If there are many docker-less users later, document will be available. Or we can support directly through installation scripts https://github.com/immich-app/immich/pull/556, so user don't need document.

Comment (panoti) [reply server/apps/immich/src/api-v1/server-info/server-info.service.ts:9]: I also think we need a documentation page to put such guides. I'm waiting for https://github.com/immich-app/immich/issues/244

**Comment by alextran1502** [comment]
After careful consideration, I think this PR fits well for your needs and setup but doesn't apply to the current vision of the main branch. I will close this PR.

**Pull request closed without merging**